src/libostree/ostree-repo-checkout.c

```diff
@@ -643,6 +647,32 @@ checkout_one_file_at (OstreeRepo                        *repo,
 
       need_copy = FALSE;
     }
+  else if (is_overlayfs_whiteout)
```

Comment:
I'm thinking we may want to make this behavior potentially opt-in?  So far ostree has no "magical filenames" - this would be the first.

A concern I have here is that e.g. flatpak would suddenly gain exposure to this.  And as we discussed it's *probably* fine...the worst I can think of is that someone using flatpak *on top of overlayfs* wouldn't be able to install an app that included a file with this name.

But still, by requiring an opt-in flag we're at least reducing the blast radius.

I think we can then hange the deploy path to pass this flag by default.  This feature would clearly be useful for everyone using ostree-for-bootable-systems.

Though I think I'd also be OK having it opt-out (i.e. we add a flag to disable this) instead; then if we discover other use cases don't want it they could turn it off.  But opt-in seems safer and only requires a little bit more work.

Reply:
I agree that opt-in as you said,may require changes in the deployment path when we install or upgrade the system. Let me talk to the RHEL4Edge team and try figure out how easy/hard it's to include the flag, probably it's easy or even under the control of the image creator so not a big deal, but I'm not familiar with that.

By looking at our composer installer image I see:
```
ostreesetup --nogpg --osname=rhel --remote=edge --url=file:///run/install/repo/ostree/repo --ref=rhel/8/x86_64/edge
```

Do you know where ostreesetup lives?


Note: It seems like any regular user with no special privileges can create char 0:0 devices , and that device always returns error.

As you said the issue would be if trying to checkout inside an overlayfs system.

Reply:
In my proposal here, the `ostree admin deploy` code would by default pass this option down when it does the checkout.  This means no modifications would be necessary for anything higher level than this project.

Conceptually ostree has two separate parts; there's a git-like core represented by `OstreeRepo`, and the sysroot/deploy bits in `OstreeSysroot`.  The split isn't 100% clean (the magical `ostree/1/0/1` type refs are part of that) but that's the idea.

Hmm actually looking at this we of course already have

```
typedef struct {
...
  gboolean process_whiteouts;
...
} OstreeRepoCheckoutAtOptions;
```

And...confusingly that's for a different model of whiteouts where we want ostree to handle them, not the container engine.  Anyways so...basically I'm saying we'd add `gboolean process_passthrough_whiteouts;` there, and then in `checkout_deployment_tree` set that boolean or so.  WDYT?

Reply:
I added an opt-out for now, but happy to change into opt-in if we see it's doable in rhel4edge or for any other reason.

Reply:
> I added an opt-out for now, but happy to change into opt-in if we see it's doable in rhel4edge or for any other reason.

oh sorry I didn't see you had written comments in the meantime, reading ^

Reply:
> 

Ok, sounds like a good plan to me

Questions to make sure I understood right:

 * so, RHEL uses ostree admin underneath ? 
 * I guess I also must verify that switch/upgrade will use the right flag too

Reply:
`ostree admin` is just a small CLI for the core `libostree` C shared library, which is also used by rpm-ostree (e.g. [here](https://github.com/coreos/rpm-ostree/blob/bc791794184802c3a405e3719f43ddbd168ae3ff/src/daemon/rpmostree-sysroot-upgrader.cxx#L1400)), which is what's used by most things in Fedora/RHEL today.  So modifying the code here will affect everything we want.

Reply:
> I guess I also must verify that switch/upgrade will use the right flag too

Basically by design, every single ostree operation ends up reconstructing a new filesystem via hardlinks - it never tries to do an "in place" update.  `checkout_deployment_tree()` is the function that does that, and everything else calls it.

Reply:
Thank you, undrestood

Reply:
ostree_sysroot_deploy_tree -> ostree_sysroot_deploy_tree_with_options -> sysroot_initialize_deployment -> checkout_deploytment_tree -> ostree_repo_checkout_at